### PR TITLE
[Enhanncement for host.ip and host.mac] Disabling netinfo.enabled option of add-host-metadata processor

### PR DESCRIPTION
**Pull request by gizas**
- Enhancement

## Proposed commit message

Relates to https://github.com/elastic/integrations/issues/6674

Please explain:

- WHAT: We are disabling the netinfo.enabled option of add-host-metadata processor by checking the existence of and environmental variable in the Elastic Agent Pod
- WHY:  Host.Ips and Host.Mac take a considerable size during indexing and affect both ingestion and visualisation process. As they dont provide meaningful infromation to the users we can disable them accordingly.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test this PR locally

1. Checkout this PR fix inside ~/beats folder of your system.
2. Clone elastic-agent repo in another folder ~/elastic-agent
3. Build elastic agent image with details provided in https://github.com/elastic/elastic-agent#testing-elastic-agent-on-kubernetes. Command `DEV=true SNAPSHOT=true PLATFORMS=linux/arm64 PACKAGES=docker mage package`
4. Build the docker image of your custom agent that you will load later in your setup: 
```bash
cd ~/elastic-agent
cd build/package/elastic-agent/elastic-agent-linux-arm64.docker/docker-build
docker build --platform linux/amd64 -t custom-agent-image .
```
5. Install kind
6. Load your custom build image into kind : `kind load docker-image custom-agent-image:latest --name kind`
7. Install your local ES stack : `elastic-package stack up -d -vvv --version=8.11.0-SNAPSHOT`
8. Use the the following manifests as samples to disable host.ip and host.mac

[elastic-agent-standalone-kubernetesHostIPs.yml.txt](https://github.com/elastic/beats/files/12523973/elastic-agent-standalone-kubernetesHostIPs.yml.txt)
[elastic-agent-standalone-kubernetesHostIPs.yml.txt](https://github.com/elastic/beats/files/12523974/elastic-agent-standalone-kubernetesHostIPs.yml.txt)

This is particular the piece of code that our manifests will need to have from now on:
```yaml
# The following NETINNFO:false variable will disable the netinfo.enabled option of add-host-metadata processor. This will remove fields host.ip and host.mac.  
            # For more info: https://www.elastic.co/guide/en/beats/metricbeat/current/add-host-metadata.html
            # - name: NETINFO
            #   value: "false"
```

## Related issues

- Relates #https://github.com/elastic/integrations/issues/6674

## Screens
Default 8.11.SNAPSHOT with host.ips and host.mac
with_8 11 SNAPSHOT
hots

Standalone Agent with NETINNFO:false
Without_hostips

Managed Agent with NETINNFO:false
Managed_withouthostips

### CODE REVIEW

**Review by ruflin — COMMENTED** [review]
* This change seems to be only done in metricbeat. Doesn't the same happen in Filebeat and any other Beat?
* Do we need additional documentation around this how a user could enable it (again) if they wanted to?

**Comment by gizas** [comment]
> * This change seems to be only done in metricbeat. Doesn't the same happen in Filebeat and any other Beat?

Indeed eg. filebeat has the same configuration see https://github.com/elastic/beats/blob/1c5541e076ebc2654ddb542e9a4ca6713ab565f2/x-pack/filebeat/cmd/root.go#L41

My first thought here is that we need it only for metricbeat and filestream (if we decide to) for now as these are related to kubernetes.

By a first look, auditbeat, osquerybeat and packetbeat initialise add_host_metadata processor

@martijnvg do you think we need it also for filebeat? Our tests were run only for metrics that is why we focused only in metricbeat I guess

> * Do we need additional documentation around this how a user could enable it (again) if they wanted to?

For now I added it only in the manifests https://github.com/elastic/elastic-agent/pull/3354/files#diff-b85244108b202e9acc2f4eca1f918bb448bbd10a816a83123ec17f7cd3ca92b6

The other place I see that might feet is here: https://www.elastic.co/guide/en/fleet/current/agent-environment-variables.html

**Comment by martijnvg** [comment]
> do you think we need it also for filebeat? Our tests were run only for metrics that is why we focused only in metricbeat I guess

I think this depends on whether local-link addresses add any value in the non metric use cases? I don't think so? But I think I'm not the person to answer this question.

**Comment by ruflin** [comment]
> By a first look, auditbeat, osquerybeat and packetbeat initialise add_host_metadata processor as well, but dont need we need it for k8s.

My understanding is that when any of these Beats / Elastic Agent are run under k8s, the problem would exist. Is there anything special here around metrics? Is the assumption that the other beats are not run in k8s?

File: x-pack/metricbeat/cmd/root.go [hunk 1]
```diff
@@ -71,10 +72,27 @@ func defaultProcessors() []mapstr.M {
...
+
+	// Setting environmental variable NETINFO:false in Elastic Agent pod will disable the netinfo.enabled option of add_host_metadata processor.
+	// This will result to events not being enhanced with host.ip and host.mac
+	// Related to https://github.com/elastic/integrations/issues/6674
+	valueNETINFO, status := os.LookupEnv("NETINFO")
+
+	if valueNETINFO == "false" && status == true {
+		return []mapstr.M{
+			{"add_host_metadata": mapstr.M{
+				"netinfo.enabled": "false",
```
Comment (ChrsMark) [inline x-pack/metricbeat/cmd/root.go:84]: Why to enable/disable this at this level and not inside the processor's implementation?
Wouldn't that be more generic without the need to configure every single Beat? And that would cover @ruflin 's concern at https://github.com/elastic/beats/pull/36506#issuecomment-1707786366, right?

Comment (felixbarny) [reply x-pack/metricbeat/cmd/root.go:84]: 👍 to do that inside the processor implementation. I think in the implementation we could even automatically check if this is running inside k8s or inside a container and either disable netinfo completely or filter out link-local addresses. See also https://github.com/elastic/integrations/issues/6674#issuecomment-1706768981.

Comment (gizas) [reply x-pack/metricbeat/cmd/root.go]: Adding the fix only to add-host-metadata processor (https://github.com/elastic/beats/pull/36506/files#diff-03b979565735e6707425275586e3215f44192f46e1ac3b456516d93c653eba13R38) seems that works only for metricbeat. Somehow seems we override it for filestream? Will keep looking

**Comment by gizas** [comment]
Lessons learned, you need to remove both builds in order to force agent image to build from local code:
`rm -rf x-pack/metricbeat/build/distributions/metricbeat-8.11.0-SNAPSHOT-linux-*`
`rm -rf x-pack/filebeat/build/distributions/filebeat-8.11.0-SNAPSHOT-linux-*`

So all work by disabling netinfo in processor : 
**Filestream:**
filestream_disabled

**Metricbeat:**
metricbeat_disabled

For documentation please review: https://github.com/elastic/ingest-docs/pull/463

File: libbeat/processors/add_host_metadata/config.go [hunk 2]
```diff
@@ -34,10 +35,24 @@ type Config struct {
...
 func defaultConfig() Config {
-	return Config{
-		NetInfoEnabled:      true,
-		CacheTTL:            5 * time.Minute,
-		ExpireUpdateTimeout: time.Second * 10,
-		ReplaceFields:       true,
+	// Setting environmental variable NETINFO:false in Elastic Agent pod will disable the netinfo.enabled option of add_host_metadata processor
+	// This will result to events not being enhanced with host.ip and host.mac
+	// Related to https://github.com/elastic/integrations/issues/6674
+	valueNETINFO, _ := os.LookupEnv("NETINFO")
```
Comment (ChrsMark) [inline libbeat/processors/add_host_metadata/config.go:41]: My only question would be the naming we want to choose here. Maybe sth like `ELASTIC_NETINFO` would be more explicit?

**Review by ChrsMark — COMMENTED** [review]
The change LGTM. I left one comment about the env variable's name.

I think that skipping by default if k8s or containarized environment is detected would possibly be a breaking change at this point. Maybe some users want these data even in k8s envs.
So I think it's safer to go with the current approach.

File: libbeat/processors/add_host_metadata/config.go (hunk 2 above)
Comment (gizas) [reply libbeat/processors/add_host_metadata/config.go:41]: I agree lets not introduce a breaking change (I was thinking an old comment you had done). 

For the naming you are killing me :) but I was expecting the debate , so seeing the page again https://www.elastic.co/guide/en/fleet/current/agent-environment-variables.html, I would go for ELASTIC_AGENT_NETINFO as strickly speaking this is a change on the agent side. 

I would wait until tomorrow to see if other have any opinion and I would need to change all prs

**Comment by ruflin** [comment]
> I think that skipping by default if k8s or containarized environment is detected would possibly be a breaking change at this point. Maybe some users want these data even in k8s envs.

I agree that this can potentially be considered a breaking change, at the same I'm not sure the use case where all these ip addresses would be used. If we don't change the default, most users will not get the benefits and have to make active changes to get the performance and storage benefits.

Do we agree `false` is the better default for most users? If we don't change it now, when would we change it? Maybe we can also separate the discussion and first get the change in to make it possible to disable and then figure out how we change the default in a good way. @felixbarny Thoughts?

File: libbeat/processors/add_host_metadata/config.go [hunk 3]
```diff
@@ -34,10 +35,24 @@ type Config struct {
...
 func defaultConfig() Config {
-	return Config{
-		NetInfoEnabled:      true,
-		CacheTTL:            5 * time.Minute,
-		ExpireUpdateTimeout: time.Second * 10,
-		ReplaceFields:       true,
+	// Setting environmental variable ELASTIC_NETINFO:false in Elastic Agent pod will disable the netinfo.enabled option of add_host_metadata processor
+	// This will result to events not being enhanced with host.ip and host.mac
+	// Related to https://github.com/elastic/integrations/issues/6674
+	valueNETINFO, _ := os.LookupEnv("ELASTIC_NETINFO")
```
Comment (ruflin) [inline libbeat/processors/add_host_metadata/config.go:41]: Lets make sure we have this documented. Can you expand the changelog to contain a mention of this env variable?

**Comment by ChrsMark** [comment]
> Do we agree false is the better default for most users?

I believe we could just set `ELASTIC_NETINFO` to `false` by default within our k8s manifests as a first step. This has 2 main advantages:
1) The codebase itself does not introduce any possible breaking change.
2) We only affect the k8s use-case but from the configuration management point of view. This means that we have a reasonable default (to `false`) only for k8s deployments which can be easily changed from users by removing the variable definition from the k8s manifest (if needed).

If we agree on this here we can even change the manifests in this PR, or change those in a follow-up if we want to discuss it more. Whatever that fits better.

**Comment by felixbarny** [comment]
SGTM. The main thing that's important to me is that `host.ip`/`host.mac` is either truncated or removed from k8s data sets **by default**.

**Comment by gizas** [comment]
> SGTM. The main thing that's important to me is that host.ip/host.mac is either truncated or removed from k8s data sets by default.

Thanks again all for the quick feedback. Default value in manifests is ELASTIC_NETINFO:false. So the host.ip and host.mac will be removed by default

**Comment by felixbarny** [comment]
> So the host.ip and host.mac will be removed by default

I that also true for logs or just for metrics?

**Comment by gizas** [comment]
> > So the host.ip and host.mac will be removed by default
> 
> I that also true for logs or just for metrics?

Both . See tests https://github.com/elastic/beats/pull/36506#issuecomment-1708304543

**Review by ChrsMark — APPROVED** [review]
The patch LGTM! Thanks!

**Review by ruflin — APPROVED** [review]
LGTM. Lets make sure we get this deployed directly in our test environments with the new configs and also get it into our cloud environments.

File: libbeat/processors/add_host_metadata/config.go [hunk 4]
```diff
@@ -34,10 +35,24 @@ type Config struct {
...
+			ExpireUpdateTimeout: time.Second * 10,
+			ReplaceFields:       true,
+		}
+	} else {
+		return Config{
+			NetInfoEnabled:      true,
+			CacheTTL:            5 * time.Minute,
+			ExpireUpdateTimeout: time.Second * 10,
+			ReplaceFields:       true,
+		}
```
Comment (felixbarny) [inline libbeat/processors/add_host_metadata/config.go:56]: Optional suggestion to increase readability

```suggestion
	return Config{
		NetInfoEnabled:      valueNETINFO == "true",
		CacheTTL:            5 * time.Minute,
		ExpireUpdateTimeout: time.Second * 10,
		ReplaceFields:       true,
	}
```

**Review by felixbarny — APPROVED** [review]

**Review by cmacknz — APPROVED** [review]

**Pull request merged**